lsコマンドを作る1の提出

maruko92188 opened the pull request:
lsコマンドを作る1の課題提出です。

maruko92188 committed:
lsコマンドを作る1の課題提出

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,30 @@
+  Dir['*'].sort_by(&:downcase)
+end
+
+output(make_output_file_names, file_names)
```
ここを起点にプログラムが始まるのであれば、main/run/execみたいな名前のメソッドがいいと思います。
https://qiita.com/jnchito/items/4b4cae54170cc2f4377e

また、ここでは引数に応じて柔軟性を持たせる必要がないので、引数はなくしてしまっていいと思います。
(mainメソッド内で呼び出す)

```suggestion
main
```

maruko92188 replied:
プログラムの起点となるメソッドをmainメソッドとして定義しました。
引数を削除しました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,30 @@
+COLUMNS = 3
+COLUMNS_DISTANCE = 5
+
+def output(make_output_file_names, file_names)
```
メソッド名と同じ引数は避けましょう
https://qiita.com/jnchito/items/fb438aa0786bb3cbf919

maruko92188 replied:
こちらはメソッド名を`output`から`main`メソッドに変更し、引数を削除したので修正はしておりません。
以後、メソッド名と同じ引数にならないよう気をつけます😃

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,30 @@
+COLUMNS_DISTANCE = 5
+
+def output(make_output_file_names, file_names)
+  longest_string = file_names.max_by(&:size)
```
stringは抽象的過ぎるので、もうちょっと具体的な名前がいいですね
https://bootcamp.fjord.jp/articles/78

maruko92188 replied:
`longest_string`と`string_width`を、それぞれ`longest_file_name`と`file_name_width`に変更しました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,30 @@
+  end
+end
+
+def make_output_file_names
```
プログラミングでは「作る」はmakeよりもbuildやcreate、generateみたいな動詞がよく使われますね

maruko92188 replied:
makeからbuildに変更します💡

JunichiIto reviewed (commented):
コメントしました！

maruko92188 committed:
起点となるメソッド名をmainに変更

maruko92188 committed:
変数名の_stringを_file_nameに変更

maruko92188 committed:
メソッド名を変更

make_output_file_nameをbuild_output_file_nameに変更した。

maruko92188 commented:
コメントありがとうございます。
ご指摘いただいた箇所を修正しました。
また、提出後に空のディレクトリで`ls.rb`を実行すると`nil`に`size`メソッドが使えないとエラーが出ることに気が付いたのですが、こちらも修正が必要でしょうか。
```
undefined method 'size' for nil (NoMethodError)
```

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,30 @@
+  sliced_file_names.transpose
+end
+
+def file_names
```
結果が同じになるにもかかわらず、このメソッドが何度も呼ばれていますね。
最近のCPUはとても速いので実感がすることはまれですが、それでも何らかの処理を行うというのはゼロコストではありません。
たとえば「ここで処理が走ってますよ」ということを意識できるように、sleepを挟んでみてください。

```suggestion
def file_names
  sleep 2
```

理想はこのsleepが1回だけ呼ばれる、つまり2秒でプログラムが完了することです。

maruko92188 replied:
`file_names`メソッドに`sleep 2`を加えてプロクラムを実行しました。出力されるのに時間がかかりました。
```ruby
def file_names
  sleep 2
  Dir['*'].sort_by(&:downcase)
end
```
Benchmarkモジュールというのを見つけたので、`build_output_file_names`の実行時間も調べて、`file_names`メソッドが3回呼ばれていることが推測できました。
```ruby
Benchmark.bm do |x|
  x.report('buiid') {build_output_file_names}
end

# =>            user     system      total        real
# => buiid  0.001309   0.001340   0.002649 (  6.017689)
```
`build_output_file_names`メソッドに`binding.break`を置いて、実際に`file_names`メソッドの中に3回入っていくのを確認しました。
```ruby
def build_output_file_names
  binding.break
  rows =  file_names.size.ceildiv(COLUMNS)
  # 省略
  sliced_file_names.transpose
end
```

maruko92188 replied:
`file_names`メソッドの返り値を変数に代入するように修正しました。
`sleep 2`を加えると、出力されるまで時間がかかりました。
修正後は2秒でメソッドを抜けるようになりました。
```ruby
# メソッドの実行にかかる時間を調べました。


# 省略

def file_names
  sleep 2
  Dir['*'].sort_by(&:downcase)
end

main

Benchmark.bm do |x|
  x.report('buiid') {build_output_file_names}
end

# 修正前
# =>            user     system      total        real
# => buiid  0.001736   0.001895   0.003631 (  6.014195)

# 修正後
# =>            user     system      total        real
# => buiid  0.000326   0.000707   0.001033 (  2.006036)
```

JunichiIto replied:
ちゃんとベンチマークを取ってるのはいいですね  👍 
ただ、対象が`build_output_file_names`だけになっています。
着目すべきはこのメソッドだけでなく、プログラム全体です。
なのでmainメソッドを対象にしてください。
もしくはtimeコマンドを使うのもありです。

```
$ time ls.rb
```

maruko92188 replied:
`file_names`の結果を引数で`main`と`build_output_file_names`に渡すように修正しました。
`file_names`に`sleep 2`を挟むと、実行時間が2秒強になりました。
```
$ time ls.rb
1fx7r4ZrFYlUY     BpeUhiLtx         qSBMpfgAr         
4Rd7              bPu               SWPLR5AQf         
8lEYP             Engayr            TqBPe             
9mzun4lC          fQmI3Yt           y45E87ehkEL       
BImEQ2            podsLfU                             
ls.rb  0.04s user 0.03s system 3% cpu 2.180 total
```

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,30 @@
+
+def main
+  longest_file_name = file_names.max_by(&:size)
+  file_name_width = longest_file_name.size + COLUMNS_DISTANCE
```
> また、提出後に空のディレクトリでls.rbを実行するとnilにsizeメソッドが使えないとエラーが出ることに気が付いたのですが、こちらも修正が必要でしょうか。

エラーが出てるのはこの行ですかね？
もしオリジナルのlsはエラーが出ないと思うので、その挙動に合わせてほしいです！

JunichiIto replied:
あと、変数名はwidthだけでもいいかもしれません
https://bootcamp.fjord.jp/pages/avoid-too-specific-var-name

maruko92188 replied:
>エラーが出てるのはこの行ですかね？

はい、こちらの行でエラーが出ていました。
`main`メソッドの始めにディレクトリが空の場合の処理を追加したところ、エラーは出なくなりました。
<img width="878" height="131" alt="blank_dir_—_-zsh_—_80×24" src="https://github.com/user-attachments/assets/a842aded-814c-4bf3-a349-11b427b1c663" />
また、widthに関係する変数が1種類のみなのでfile_name_widthをwidthだけにしました。

JunichiIto reviewed (commented):
コメントしました〜。

maruko92188 committed:
build_output_file_namesメソッドの動きを改善

file_namesメソッドが何度も呼ばれないように、始めに変数に代入するようにした。

maruko92188 committed:
空のディレクトリで実行してもエラーが出ないように修正

maruko92188 committed:
file_name_widthをwidthに変更

maruko92188 committed:
file_namesメソッドの結果を引数として渡すように修正

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,31 @@
+  Dir['*'].sort_by(&:downcase)
+end
+
+main(file_names)
```
mainに引数を渡してしまうと、下のコメントにまた戻ってきてしまいますよ？
https://github.com/maruko92188/ruby-practices/pull/4#discussion_r2400515832

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,31 @@
+COLUMNS = 3
+COLUMNS_DISTANCE = 5
+
+def main(file_names)
```
ここもまた戻ってきてますね。。
https://github.com/maruko92188/ruby-practices/pull/4#discussion_r2400517317

maruko92188 replied:
main内でfile_namesが呼ばれるように修正しました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,31 @@
+COLUMNS_DISTANCE = 5
+
+def main(file_names)
+  return exit if file_names.empty?
```
exitは呼んだ瞬間にプログラムが終了するのでreturnする必要がないです。

```suggestion
  exit if file_names.empty?
```

ただし、exitは基本的に「プログラムの緊急停止ボタン」なので、「どうしても」というとき意外は使わない方がいいです。
ここはexitを使わずにただのreturnにすれば、処理を中断することができます。


```suggestion
  return if file_names.empty?
```

maruko92188 replied:
exitを削除しました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,31 @@
+def main(file_names)
+  return exit if file_names.empty?
+  longest_file_name = file_names.max_by(&:size)
+  width = longest_file_name.size + COLUMNS_DISTANCE
```
上の画像はオリジナルのlsと列幅が揃っていますが、下の方が揃ってないですね。

<img width="704" height="769" alt="Image" src="https://github.com/user-attachments/assets/c1829454-f2a9-44c4-afc2-41dcebace530" />

何か規則性があってそれを再現できそうならがんばって再現してほしいですが、難しそうなら単純に

```suggestion
  width = longest_file_name.size + 1
```

だけでもOKです。
（`COLUMNS_DISTANCE = 5`だと、あるときは揃うが、あるときは揃わない中途半端な状態になるので）

maruko92188 replied:
規則性は見つけられませんでいたが、出力されるファイル名の後ろにタブを挿入するとオリジナルのlsと列幅が揃いました。
ただ、オリジナルのlsコマンドでは列の間はタブしかなく、最終列の後ろには何も出力されていませんでした。
こちらを再現するのは難しそうです💦
<img width="816" height="465" alt="_marukokazuki___Desktop_test__•_Untitled-1" src="https://github.com/user-attachments/assets/0a3717f8-bdbf-416a-a8b7-f67bad214c8b" />

JunichiIto replied:
> ただ、オリジナルのlsコマンドでは列の間はタブしかなく、最終列の後ろには何も出力されていませんでした。こちらを再現するのは難しそうです💦

いったん配列に書く文字列を放り込んでタブ文字でjoinすると、再現できるかも？

JunichiIto reviewed (commented):
コメントしました！

maruko92188 committed:
file_namesがmain内で1度だけ呼ばれるように修正

maruko92188 committed:
exitを使用しないように変更

maruko92188 committed:
列の間にタブが入るように変更

maruko92188 committed:
widthの算出を1つの式にまとめた

maruko92188 committed:
rubocopの指摘箇所を修正

main内のガード節の後に空行を追加
main内のファイル名出力を式展開にした

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,31 @@
+COLUMNS = 3
+
+def main
+  created_file_names = file_names
```
```suggestion
  file_names = file_names
```

変数名の方をシンプルにしたいですね。ということは・・・？
ヒント https://qiita.com/jnchito/items/fb438aa0786bb3cbf919

maruko92188 replied:
メソッド名を`create_file_names`に変更して変数名をシンプルにしました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,31 @@
+  end
+end
+
+def build_output_file_names(created_file_names)
```
配列の配列が戻り値になるので、単純な複数形にするより

- grid
- table
- matrix
- nested_something

みたいな単語を使って、「配列の配列が返りますよ」ということをアピールしたいですね。
（単純な複数形だとフラットな配列が返るように見える）

maruko92188 replied:
tableを使ったメソッド名に変更しました。

JunichiIto reviewed (commented):
コメントしました！

maruko92188 committed:
メソッド名と変数名を修正

メソッド名`file_names`を`create_file_names`に変更
メソッド名`build_output_file_names`を`build_file_names_table`に変更
変数`created_file_names = file_names`を`file_names = create_file_names`に変更

maruko92188 committed:
列の間をljustではなく、タブ文字で空けるように変更

maruko92188 committed:
出力するファイル名に``がついていたのを削除

maruko92188 commented:
列幅について、アドバイスしていただいた`文字列を配列に入れてからタブ文字でjoinする`という方法を考えましたが、『表示するファイル名ごとに、必要なタブ文字の数が変わる部分をどうするか』というところが分からず行き詰まってしまいました。
（joinの引数にわたすタブ文字の数をどう変えたら良いかがわかりませんでした。）

そこで、ファイル名が表示されるごとに、タブ文字が必要数加えられるという方法で出力するようにしました。
<img width="659" height="195" alt="test_—_-zsh_—_60×25" src="https://github.com/user-attachments/assets/472b1e5b-fe92-426e-9daa-695c2be2d88e" />
```
$ ls   
7dQ3		Iy5K8gY7	ppo8hnLn
ficsdm		jptddZDEAGJ	U5Wnns

$ ls   \n7dQ3\t\tIy5K8gY7\tppo8hnLn\nficsdm\t\tjptddZDEAGJ\tU5Wnns\n
```
```
$ ls.rb
7dQ3		Iy5K8gY7	ppo8hnLn
ficsdm		jptddZDEAGJ	U5Wnns

$ ls.rb\n7dQ3\t\tIy5K8gY7\tppo8hnLn\nficsdm\t\tjptddZDEAGJ\tU5Wnns\n
```

maruko92188 commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,34 @@
+    last_index = row.size
+    row.each.with_index(1) do |file_name, index|
+      tab_count = longest_file_name.div(TAB_SPACE) + 1 - file_name.to_s.size.div(TAB_SPACE)
+      print index == last_index ? file_name : "#{file_name}#{"\t" * tab_count}"
```
この部分を以下のように書いてエラーが出たのですが、エラーの解釈が間違っていないかを確認したいです。
```ruby
 index == last_index ? print "#{file_name}" : print "#{file_name}#{"\t" * tab_count}"
```
```
/Users/marukokazuki/bin/ls.rb:16: syntax errors found (SyntaxError)
  14 |     row.each.with_index(1) do |file_name, index|
  15 |       tab_count = longest_file_name.div(TAB_SPACE) + 1 - file_name.to_s.size.div(TAB_SPACE)
> 16 | ...  "#{file_name}" : print "#{file_name}#{"\t" * t ...
     |     ^ expected a `:` after the true expression of a ternary operator
     |                     ^ unexpected ':', ignoring it
     |                     ^ unexpected ':', expecting end-of-input
```

maruko92188 replied:
`条件式 ? 真の場合の処理 : 偽の場合の処理`で、真の場合の処理が`?` の直後の`print`のみと認識されているのでエラーが出ていると解釈しました。
そこで、真の場合の処理と偽の場合の処理を明確にするため、以下のように書き換えるとエラーは解消されました。
```ruby
index == last_index ? (print "#{file_name}") : (print "#{file_name}#{"\t" * tab_count}")
```
最終的には`print`は条件分岐の前にしたのですが、エラーの解釈が合っていたのか確認したいです🙇

JunichiIto replied:
```ruby
index == last_index ? (print "#{file_name}") : (print "#{file_name}#{"\t" * tab_count}")
```

たしかに↑でもエラーはなくなりますが、それよりもprintの引数を()で囲む方が一般的かも。

```ruby
index == last_index ? print("#{file_name}") : print("#{file_name}#{"\t" * tab_count}")
```

> 最終的にはprintは条件分岐の前にした

それもありですが、あまり1行に処理を詰め込まず、いったん変数に入れる方がいいかもです。

```ruby
foo = index == last_columns_number ? file_name : "#{file_name}#{"\t" * tab_count}"
print foo
```

> エラーの解釈が合っていたのか確認したいです🙇

はい、解釈は合ってそうです。
（expected a `:` after the true expression of a ternary operator のあたりが該当しそう）

ちなみに求められてる回答とは異なるかもしれませんが、Rubyはメソッド呼び出しの`()`を省略できるとはいえ、いろんな記法を組み合わせると、書き手の意図した通り解釈されないことがよくあります。
なので、今回のようなエラーが起きた場合は「しまった、ちょっと()を省略しすぎたか」と思って、`foo x`を`foo(x)`に修正しますね。（で、それでもダメなら別の原因を探る）

maruko92188 replied:
回答ありがとうございます！
`print x`や`puts x` は`()`が省略されているという認識が定着してませんでした。（チェリー本を見返したら書いてました😅）
>Rubyはメソッド呼び出しの()を省略できるとはいえ、いろんな記法を組み合わせると、書き手の意図した通り解釈されないことがよくあります

エラーの対処の方法の1つとして覚えておきます！

maruko92188 committed:
変数名last_indexをlast_columns_numberに変更

maruko92188 committed:
定数名TAB_SPACEをTAB_WIDTHに変更

maruko92188 committed:
列幅の計算をループ処理の前に行うように変更

maruko92188 committed:
column_widthを求める処理をメソッド化

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,30 +2,38 @@
+    last_columns_number = row.size
+    row.each.with_index(1) do |file_name, index|
+      tab_count = column_width - file_name.to_s.size.div(TAB_WIDTH)
+      print index == last_columns_number ? file_name : "#{file_name}#{"\t" * tab_count}"
```
> 列幅について、アドバイスしていただいた文字列を配列に入れてからタブ文字でjoinするという方法を考えましたが、『表示するファイル名ごとに、必要なタブ文字の数が変わる部分をどうするか』というところが分からず行き詰まってしまいました。
（joinの引数にわたすタブ文字の数をどう変えたら良いかがわかりませんでした。）

あー、たしかにタブ数が動的に変わる場合はjoinは使えないですね。

であればいったんどの列もタブ文字付きで配列に入れて、最後にjoinしてrstripする（＝行末のタブ文字を消す）、みたいな方法がいいかもしれません。

```ruby
puts columns.join.rstrip
```

maruko92188 replied:
ファイル名を取得してすぐに、タブ文字を加えるように処理の順番を変更しました。
`rstrip`👀このメソッドで『最終列の後ろの余計なタブ文字をどうしよう❓』という悩みが解決しました💡
出力の条件分岐も必要なくなり、mainメソッドがスッキリしました。
便利なメソッドが色々用意されていますね😄

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -2,30 +2,38 @@
 end
 
-def file_names
+def create_file_names
```
createより、OSからファイルの一覧を探し出してくる、みたいなイメージでsearchやfindを使う方がいいかも。

maruko92188 replied:
createからsearchに変更しました。

JunichiIto reviewed (commented):
コメントしました！

maruko92188 committed:
メソッド名create_file_namesをsearch_file_namesに変更

maruko92188 committed:
処理の流れを変更

ファイル名の取得後にタブ文字を加えるようにした
この処理を行うメソッドadd_tab_characterを新たに定義
メソッドの順を呼ばれる順に変更した

maruko92188 committed:
search_file_namesメソッドを定義する位置を変更

maruko92188 committed:
add_tab_characterをadd_tab_charactersに変更
